### PR TITLE
usb_device_check: add check max_devices_under_one_hub and mouse/tablet/kbd cases

### PR DESCRIPTION
**Pull request by hereischen**
Add max_devices_under_one_hub, to check 8 usb mouse attatced to one usb hub.
Add usb_mouse, to check a usb mouse attached without usb hub.
usb_mouse_and_tablet checks a usb mouse and tablet without usb hub.
usb_kbd checks a usb keyboard without usb hub.
usb_common: modified imported modules to make it compatible with both python 2.6 and 2.7

id: 1493404 1493413 1493420 1493423 1494469

Signed-off-by: Haotong Chen hachen@redhat.com

### CODE REVIEW

**Comment by hereischen** [comment]
PASS 1-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.max_devices_under_one_hub.nec-xhci

PASS 1-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_mouse.usb-ehci

@ngu-niny  Please help to review, thanks.

File: qemu/tests/cfg/usb_device_check.cfg [hunk 1]
```diff
@@ -85,3 +85,18 @@
         - input_devices_under_two_tier_hub:
             no usb-ehci
             usb_topology = '{"usb-hub":2,"usb-mouse":1,"usb-kbd":1,"usb-tablet":1}'
+        - max_devices_under_one_hub:
+            no usb-ehci
+            usb_topology = '{"usb-hub":1,"usb-mouse":8}'
+        - usb_mouse:
```
Comment (CongLi) [inline qemu/tests/cfg/usb_device_check.cfg]: For this scenario, there is step for checking dmesg info of usb device, could you please show me it? I have not found it.

Thanks.

Comment (hereischen) [reply qemu/tests/cfg/usb_device_check.cfg]: I'd consider they are same. So I only use `lsusb` here.

**Comment by ngu-niny** [comment]
Test PASS:

'01-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.max_devices_under_one_hub.piix3-uhci' Finshed 2017-09-21 03:22:41 2017-09-21 03:29:15 PASS 394.599456
'02-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.max_devices_under_one_hub.piix4-uhci' Finshed 2017-09-21 03:29:16 2017-09-21 03:29:58 PASS 42.329423
'03-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.max_devices_under_one_hub.ich9-uhci' Finshed 2017-09-21 03:29:58 2017-09-21 03:30:42 PASS 43.565278
'04-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.max_devices_under_one_hub.ich9-ehci' Finshed 2017-09-21 03:30:42 2017-09-21 03:31:25 PASS 42.816076
'05-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.max_devices_under_one_hub.nec-xhci' Finacceptanceshed 2017-09-21 03:31:25 2017-09-21 03:32:07 PASS 42.228062
'06-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_mouse.piix3-uhci' Finshed 2017-09-21 03:32:07 2017-09-21 03:32:50 PASS 42.437037
'07-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_mouse.piix4-uhci' Finshed 2017-09-21 03:32:50 2017-09-21 03:33:33 PASS 42.770995
'08-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_mouse.ich9-uhci' Finshed 2017-09-21 03:33:33 2017-09-21 03:34:15 PASS 42.047094
'09-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_mouse.ich9-ehci' Finshed 2017-09-21 03:34:15 2017-09-21 03:34:58 PASS 42.573679
'10-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_mouse.usb-ehci' Finshed 2017-09-21 03:34:58 2017-09-21 03:35:40 PASS 42.553206
'11-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_mouse.nec-xhci' Finshed 2017-09-21 03:35:40 2017-09-21 03:36:23 PASS 42.267679
'12-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_mouse_and_tablet.piix3-uhci' Finshed 2017-09-21 03:36:23 2017-09-21 03:37:05 PASS 42.55882
'13-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_mouse_and_tablet.piix4-uhci' Finshed 2017-09-21 03:37:06 2017-09-21 03:37:48 PASS 42.115187
'14-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_mouse_and_tablet.ich9-uhci' Finshed 2017-09-21 03:37:48 2017-09-21 03:38:30 PASS 42.306068
'15-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_mouse_and_tablet.ich9-ehci' Finshed 2017-09-21 03:38:30 2017-09-21 03:39:13 PASS 42.352542
'16-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_mouse_and_tablet.usb-ehci' Finshed 2017-09-21 03:39:13 2017-09-21 03:39:55 PASS 42.207308
'17-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_mouse_and_tablet.nec-xhci' Finshed 2017-09-21 03:39:55 2017-09-21 03:40:38 PASS 42.201408
'18-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_kbd.piix3-uhci' Finshed 2017-09-21 03:40:38 2017-09-21 03:41:19 PASS 41.709349
'19-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_kbd.piix4-uhci' Finshed 2017-09-21 03:41:20 2017-09-21 03:42:03 PASS 43.022888
'20-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_kbd.ich9-uhci' Finshed 2017-09-21 03:42:03 2017-09-21 03:42:45 PASS 41.705074
'21-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_kbd.ich9-ehci' Finshed 2017-09-21 03:42:45 2017-09-21 03:43:28 PASS 42.551108
'22-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_kbd.usb-ehci' Finshed 2017-09-21 03:43:28 2017-09-21 03:44:10 PASS 42.631575
'23-smp_2.2048m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_de  vice_check.usb_kbd.nec-xhci' Finshed 2017-09-21 03:44:11 2017-09-21 03:44:53 PASS 42.660954

Acked-by: Nini Gu <ngu@redhat.com>

**Comment by hereischen** [comment]
PASS 1-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_kbd.usb-ehci

File: qemu/tests/cfg/usb_device_check.cfg [hunk 2]
```diff
@@ -85,3 +86,18 @@
...
+            Host_RHEL.m6:
+                no usb-ehci
+            usb_topology = '{"usb-mouse":1}'
+        - usb_mouse_and_tablet:
+            Host_RHEL.m6:
+                no usb-ehci
+            usb_topology = '{"usb-mouse":1,"usb-tablet":1}'
+        - usb_kbd:
+            Host_RHEL.m6:
```
Comment (ngu-niny) [inline qemu/tests/cfg/usb_device_check.cfg]: @hereischen There is a bug on rhel6, would you also want to fix it in the patch?

Comment (hereischen) [reply qemu/tests/cfg/usb_device_check.cfg]: updated

**Comment by hereischen** [comment]
on rhel6, python 2.6 :
PASS 1-smp_4.8192m.repeat1.run_test.Host_RHEL.m6.u9.qcow2.virtio_scsi.up.virtio_net.RHEL.6.9.x86_64.io-github-autotest-qemu.usb_device_check.usb_kbd.ich9-ehci

PASS 2-smp_4.8192m.repeat1.run_test.Host_RHEL.m6.u9.qcow2.virtio_scsi.up.virtio_net.RHEL.6.9.x86_64.io-github-autotest-qemu.usb_device_check.input_devices_under_two_tier_hub.piix4-uhci

on rhel7,python 2.7:
PASS 1-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.input_devices_under_two_tier_hub.nec-xhci
PASS 1-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_kbd.nec-xhci

**Comment by ngu-niny** [comment]
Test PASS.

On RHEL7:
'01-smp_8.1024m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.hub_speed_mismatch.usb-ehci' Finshed 2017-10-10 00:39:52 2017-10-10 00:39:57 PASS 4.637126
'02-smp_8.1024m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.multi_hubs_in_series.5_layer.piix3-uhci' Finshed 2017-10-10 00:39:57 2017-10-10 00:46:37 PASS 399.804255
'13-smp_8.1024m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.multi_hubs_in_parallel.piix4-uhci' Finshed 2017-10-10 00:51:10 2017-10-10 00:52:01 PASS 51.341767
'19-smp_8.1024m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.input_devices_under_one_hub.ich9-uhci' Finshed 2017-10-10 00:56:17 2017-10-10 00:57:07 PASS 49.727291
'31-smp_8.1024m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.max_devices_under_one_hub.nec-xhci' Finshed 2017-10-10 01:07:46 2017-10-10 01:08:45 PASS 58.807617
'32-smp_8.1024m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_mouse.piix3-uhci' Finshed 2017-10-10 01:08:45 2017-10-10 01:09:36 PASS 50.853009
'41-smp_8.1024m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_mouse_and_tablet.ich9-ehci' Finshed 2017-10-10 01:16:18 2017-10-10 01:17:06 PASS 48.809409
'49-smp_8.1024m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_kbd.nec-xhci' Finshed 2017-10-10 01:23:00 2017-10-10 01:23:50 PASS 50.166382

On RHEL6:
'01-smp_4.2048m.repeat1.run_test.Host_RHEL.m6.u9.qcow2.virtio_scsi.up.virtio_net.RHEL.6.9.x86_64.io-github-autotest-qemu.usb_device_check.hub_speed_mismatch.usb-ehci' Finshed 2017-10-10 02:07:04 2017-10-10 02:07:07 PASS 3.62362
'04-smp_4.2048m.repeat1.run_test.Host_RHEL.m6.u9.qcow2.virtio_scsi.up.virtio_net.RHEL.6.9.x86_64.io-github-autotest-qemu.usb_device_check.multi_hubs_in_series.5_layer.ich9-ehci' Finshed 2017-10-10 02:09:13 2017-10-10 02:10:16 PASS 62.876397
'06-smp_4.2048m.repeat1.run_test.Host_RHEL.m6.u9.qcow2.virtio_scsi.up.virtio_net.RHEL.6.9.x86_64.io-github-autotest-qemu.usb_device_check.multi_hubs_in_series.6_layer.piix4-uhci' Finshed 2017-10-10 02:10:20 2017-10-10 02:10:23 PASS 3.519324
'10-smp_4.2048m.repeat1.run_test.Host_RHEL.m6.u9.qcow2.virtio_scsi.up.virtio_net.RHEL.6.9.x86_64.io-github-autotest-qemu.usb_device_check.multi_hubs_in_parallel.ich9-ehci' Finshed 2017-10-10 02:12:38 2017-10-10 02:13:43 PASS 65.06033
'13-smp_4.2048m.repeat1.run_test.Host_RHEL.m6.u9.qcow2.virtio_scsi.up.virtio_net.RHEL.6.9.x86_64.io-github-autotest-qemu.usb_device_check.input_devices_under_one_hub.ich9-ehci' Finshed 2017-10-10 02:15:50 2017-10-10 02:16:48 PASS 58.572757
'15-smp_4.2048m.repeat1.run_test.Host_RHEL.m6.u9.qcow2.virtio_scsi.up.virtio_net.RHEL.6.9.x86_64.io-github-autotest-qemu.usb_device_check.input_devices_under_two_tier_hub.piix4-uhci' Finshed 2017-10-10 02:17:49 2017-10-10 02:18:49 PASS 60.24283
'17-smp_4.2048m.repeat1.run_test.Host_RHEL.m6.u9.qcow2.virtio_scsi.up.virtio_net.RHEL.6.9.x86_64.io-github-autotest-qemu.usb_device_check.max_devices_under_one_hub.piix3-uhci' Finshed 2017-10-10 02:19:50 2017-10-10 02:20:54 PASS 63.318815
'20-smp_4.2048m.repeat1.run_test.Host_RHEL.m6.u9.qcow2.virtio_scsi.up.virtio_net.RHEL.6.9.x86_64.io-github-autotest-qemu.usb_device_check.usb_mouse.piix3-uhci' Finshed 2017-10-10 02:22:58 2017-10-10 02:24:00 PASS 62.306153
'24-smp_4.2048m.repeat1.run_test.Host_RHEL.m6.u9.qcow2.virtio_scsi.up.virtio_net.RHEL.6.9.x86_64.io-github-autotest-qemu.usb_device_check.usb_mouse_and_tablet.piix4-uhci' Finshed 2017-10-10 02:27:09 2017-10-10 02:28:10 PASS 60.994886
''28-smp_4.2048m.repeat1.run_test.Host_RHEL.m6.u9.qcow2.virtio_scsi.up.virtio_net.RHEL.6.9.x86_64.io-github-autotest-qemu.usb_device_check.usb_kbd.ich9-ehci' Finshed 2017-10-10 02:31:18 2017-10-10 02:32:20 PASS 61.672789

Acked-by: Nini Gu <ngu@redhat.com>

**Comment by hereischen** [comment]
@luckyh  do you have any comments?

File: qemu/tests/usb_device_check.py [hunk 3]
```diff
@@ -49,7 +49,16 @@ def _check_test_step_result(result, output):
 
     error_context.context("verify usb devices information in guest...",
                           logging.info)
-    result, output = verify_usb_device_in_guest(params, session, devs)
+    # using dmesg, Linux only
+    result, output = verify_usb_device_in_guest(params, session, devs,
+                                                params["chk_dmesg"])
```
Comment (luckyh) [inline qemu/tests/usb_device_check.py]: `chk_dmesg = "dmesg| grep usb"` this is too general to check the given usb devices, and also I think it is not necessary to verify the usb devices by checking dmesg log.

Comment (hereischen) [reply qemu/tests/usb_device_check.py]: So do you mean I change it back?

Comment (luckyh) [reply qemu/tests/usb_device_check.py]: Yep, thanks.

Comment (hereischen) [reply qemu/tests/usb_device_check.py]: updated thanks

**Comment by luckyh** [comment]
The code looks good to me.

@hereischen can you please provide the test result?

**Comment by hereischen** [comment]
PASS 1-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.usb_device_check.usb_mouse_and_tablet.usb-ehci

**Comment by luckyh** [comment]
Thanks all.

**Pull request merged**